Revamp Chatbot Prototype

sqr00t opened the pull request:
---

# Description

Revamps the NIBE prototype to use latest Langchain. Also separates responsibilities:

- asf_hp_installer_chatbot: *experimentation* of data heavy pipelines, with associated utils
- rag: all things end-to-end RAG Langchain related, this includes document processing pipelines
- api: uses Langserve, a FastAPI wrapper. This enables more ways to interact with the chains throughout dev, test, and prod.

Note: WhatsApp interface hasn't been implemented, and this PR will be updated to reflect these changes. The aim of this PR is to first revamp the prototype to use the latest Langchain LCEL framework.

Addresses #6 , branch is `2_revamp`

# Instructions for Reviewer

## **Contact me for Langfuse integration details (env vars needed in your `.env` file)**

1. Checkout branch `2_revamp`
2. Run `make pip-install` if you already have the project conda environment setup
3. Test the integration test script that mimics the prototype test script:

```bash
python rag/tests/test_retrieval.py
```

To test the API:
1. Have `poetry` installed, feel free to ask me how best to do this
2. `conda deactivate`, then `cd api`
3. `poetry env use 3.10`, then `poetry install`
4. `langchain app serve --app app.server:app`
5. Navigate to `0.0.0.0:8000`
6. Test out the `/invoke` endpoint by replacing the string in the JSON request `"input": "string"`, with a query.

Please pay special attention to and other notes

- Setting up, feel free to ask me questions
- I'm still rethinking/ implementing the WhatsApp interface and the messaging system, will be best to not serialise the files but to log conversations to a message queuing system i.e. Redis. Then periodically send incremental batches to an S3 parquet file.

# Checklist:

- [x] I have refactored my code out from `notebooks/` N/A
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`

~~- [ ] I have documented the code~~ To be addressed in follow up documentation issue/ PR.
  ~~- [ ] Major functions have docstrings~~ To be addressed in follow up documentation issue/ PR.
  ~~- [ ] Appropriate information has been added to `README`s~~ To be addressed in follow up documentation issue/ PR.

- [x] I have explained this PR above
- [x] I have requested a code review

# Other additional features to review:

- [x] Add CLI for running ingestion and reindexing of docs into vector database

~~- [ ] Add CLI command for saving vector database snapshots to local and desired S3 path~~ Nice to have. Not implemented at this stage.
~~- [ ] Add CLI command for recreating an index from snapshot saved locally or in S3 path~~ Nice to have. Not implemented at this stage.

- [x] Add ability to not need to recreate local vector databases without hosting a local Docker DB instance

Review instructions, after `make install` or `make pip-install`, test out CLI commands with `chatbot --help`, there are descriptions for other commands available.

helloaidank committed:
small changes

helloaidank committed:
modularise code for the first prototype of the chatbot

helloaidank committed:
keep a directory and change doc string in chatbot script

helloaidank committed:
change in a few doc strings of testing script

sqr00t committed:
deps: update langchain dependencies

sqr00t committed:
build: bump to python3.10, use nodefaults conda channel

sqr00t committed:
build(pre-commit): add nbstripout

sqr00t committed:
feat(revamp): add utils and scripts using new langchain LCEL API

sqr00t committed:
docs(LCEL example): add example notebook on using LCEL with hp_installer_chatbot chains

sqr00t committed:
docs(LCEL example): add chain dag images

sqr00t committed:
deps(api): prevent setuptools from installing api directory

sqr00t committed:
test(revamp): add retrieval integration test that mirrors prototype test

helloaidank committed:
changes suggested by Roisin

sqr00t committed:
feat(api): add poetry project structure, scripts for API app, dependencies, and docs

helloaidank committed:
suggested changes around configuration and re-jigging functions

helloaidank committed:
small change to yaml file

helloaidank committed:
change docstring

sqr00t committed:
build(envs): improve env vars loading, add env template

sqr00t committed:
feat(hosting): add ngrok tunnel capability

sqr00t committed:
feat(chat memory): add chat memory history chains, templates, and utils

sqr00t committed:
feat(whatsapp): add webhook endpoint for chatbot with chat memory

sqr00t committed:
refactor(vectordb config): add config for hosted remote qdrant service

sqr00t committed:
fix(document loading): add preprocessing to conform to json validation

sqr00t committed:
deps(api): update api dev dependencies

sqr00t committed:
build: add deployment docker-compose config and deps, move .dockerignore

sqr00t committed:
docs(hosting): add deployment steps, update .env.template

sqr00t committed:
fix: merge dev and resolve conflicts

sqr00t committed:
chore: add qdrant data folder to .gitignore

sqr00t committed:
small refactor(llm): update model config defaults to gpt-4o

sqr00t committed:
build: cleanup repo config, add mappings to config, add cli entry point to setup.py

sqr00t committed:
feat(doc loading): add directory loading for multiple pdfs, update source mapping for docs, increase use of base config

sqr00t committed:
feat(cli): add commands for initialising vectory database and ingesting

sqr00t committed:
feat(expand vdb): update prompts to account for more docs in knowledge base

sqr00t committed:
deps: fix compatibility issues with unstructured

sqr00t committed:
docs(LCEL example): add examples for retrieving generation traces from langfuse

helloaidank commented on api/Dockerfile:
Just for my understanding, we are packaging our application and all its dependencies into a Docker image and then running this image on an EC2 instance. This has the advantage of not having to worry about the underlying OS configuration?

sqr00t replied:
Yes, to that effect. It also applies some bug fixes and some better practices for caching when building Docker images.

The lines creating a non-privileged user have been temporarily commented out temporarily.

helloaidank commented on api/app/client.py:
```diff
@@ -0,0 +1,60 @@
+
+
+async def chat(request: Request) -> Response:
+    """Interface for the Twilio WhatsApp webhook.
```
Potentially just expanding this doc string a bit so it's clearer what this function actually does.
Something like:
"This function processes incoming WhatsApp messages from Twilio, sends them to the RAG chatbot for processing, and sends the chatbot's responses back to Twilio."

helloaidank commented on api/app/client.py:
```diff
@@ -0,0 +1,60 @@
+    response = MessagingResponse()
+    response.message(chain_res["answer"])
+
+    return Response(content=str(response), media_type="application/xml")
```
Just wondering why use the xml format for the responses?

sqr00t replied:
I was following [this example](https://www.twilio.com/en-us/blog/build-secure-twilio-webhook-python-fastapi)

Strictly speaking, Twilio uses TwiML, a modified XML.

helloaidank commented on api/app/server.py:
```diff
@@ -0,0 +1,59 @@
+
+
+@app.get("/")
+async def redirect_root_to_docs():
```
What's this function doing?

sqr00t replied:
Redirects the "/" endpoint, also known as the root endpoint to the OpenAPI swagger documentations specification (not to be confused with OpenAI).

The function name is self descriptive, so I'd usually omit docstrings

helloaidank commented on api/app/server.py:
```diff
@@ -0,0 +1,59 @@
+
+app.add_route("/hook", chat, methods=["POST"])
+
+chatbot_chain = rag_chain_with_source.with_config(
```
Not too sure what's this doing?

sqr00t replied:
currently for ease, although there's a separate "whatsapp interface" client.py file, the whatsapp client and the chatbot chains api scripts are hosted on the same instance.

This adds the /hook endpoint as a POST method.

with_config, adds the langfuse_handler that allows logging instances that use this chain

The separate add_routes() methods a few lines below is a method from LangServe that abstracts away adding schema, and the extra endpoints and processing needed for /stream /batch and other endpoints.

helloaidank commented on api/app/utils/events.py:
```diff
@@ -0,0 +1,47 @@
+
+    Args:
+        app (FastAPI): The FastAPI application.
+    """
```
`" Yields:
        AsyncGenerator[Any, None]: This function is a context manager, so it doesn't return a value. Instead, it sets up the context for the FastAPI application to run, then yields control back to the application. When the application is done, it resumes to perform cleanup tasks."`

helloaidank replied:
Potentially add this to the doc string for clarity

helloaidank commented on api/pyproject.toml:
```diff
@@ -0,0 +1,42 @@
+[tool.poetry]
```
Potentially just a quick explanation on the difference between this `pyproject.toml` and the `build.toml` file for clarity?

sqr00t replied:
build.toml is used when building the Docker image (see Dockerfile line 24)

- the main difference is that the `asf-hp-installer-chatbot` package is able to be linked via relative paths (line 37 of pyproject.toml)
- when building the image for linux/amd64 architectures, to avoid downloading and compiling CUDA dependencies that will result in 5x the image size, the link to the binaries are included directly (line 33, and 34 of build.toml)
- also when building the image, the source dependencies from `asf-hp-installer-chatbot` are directly copied, therefore included separately.

helloaidank commented on rag/chains.py:
```diff
@@ -0,0 +1,34 @@
+
+rag_with_history_chain = chatbot_with_history_template | openai_llm | StrOutputParser()
+
+chatbot_with_history_chain = rag_with_source_base.assign(
```
is my understanding correct that the RAG chain with History will add the history to the prompt (i.e. we have the prompt + messaging history)?

sqr00t replied:
there is a basic implementation of RAG with in-memory message history. it would better to use a Redis implementation but this is a nice to have/ not in the current scope.

sqr00t committed:
build(deps): remove source reference in torch and torchvision

helloaidank commented on rag/prompt_templates.py:
```diff
@@ -0,0 +1,82 @@
+
+"""
+
+chatbot_system_message = SystemMessage(
```
I think something we will want to add in here is sticking to UK english as this was feedback we got from some user testing with a domain expert!

sqr00t replied:
Interesting, was there a record of specifically what they tried asking in the different language?

helloaidank replied:
No, no, it's because the chatbot uses american english and we are dealing with a UK audience. One of the users found this grating!

helloaidank replied:
Given gpt-4o-mini looks quite verbose, I think it would be great to add in a line here like "Please be concise where possible"

helloaidank commented on rag/prompt_templates.py:
```diff
@@ -0,0 +1,82 @@
+    MessagesPlaceholder,
+)
+
+optional_preprompt = """
```
this prompt is probably superseded by the below prompt? Or is the idea to keep it as it's a much more slimmed down version of the below. If so then we should update the extent of the knowledge base (which we will be updating to satisfy Renbees use)

sqr00t replied:
This is the old "system" prompt. In the earlier implementation of the chatbot, the prompt was the pre-prompt concatenated with the query, there was no system prompt used.

sqr00t replied:
In effect, I've kept the old prompt for reuse, for example, in an integration test script

helloaidank commented on rag/utils/data_loaders.py:
```diff
@@ -0,0 +1,27 @@
+    Args:
+        doc_directory (Union[Path, str], optional): Directory containing heatpump manuals and guides. Defaults to `config["doc_dir"]`.
+    Returns:
+        List[Document]: Heatpump Manuals and Guides as chunked Langchain Documents.
```
```suggestion
        List[Document]: Heat pump Manuals and Guides as chunked Langchain Documents.
```

helloaidank commented on rag/vector_databases/databases.py:
```diff
@@ -0,0 +1,101 @@
+import os
```
Think you forgot to put doc strings on these functions!

helloaidank commented on rag/vector_databases/retrievers.py:
```diff
@@ -0,0 +1,14 @@
+from rag.vector_databases import init_vdb
```
This a broader reflection, but it would be very nice if at some point in our development we try out a couple of 'advanced' retrieval techniques.

There are two options which I've learned about and think are pretty neat, which appear to be only to be available with LlamaIndex which is [sentence window retrieval](https://medium.com/generative-ai/advanced-rag-retrieval-strategies-sentence-window-retrieval-b6964b6e56f7) and [auto-merging retrieval](https://medium.com/generative-ai/advanced-rag-retrieval-strategies-auto-merging-retrieval-dc3f869654c4).

Another option which can be integrated with LangChain is neo4j, which enables us to use [more advanced retrieval techniques using Knowledge Graphs](https://medium.com/neo4j/implementing-advanced-retrieval-rag-strategies-with-neo4j-c968a002c513)! 

It would be a super nice output to evaluate different retrieval techniques and how they score for metrics such as answer relevance, faithfulness and context relevance.

sqr00t replied:
Llamaindex retrieving functions can be composed and wrapped into a function to call in a Langchain workflow. Let's put these as nice to haves to work on in a future sprint!

helloaidank commented on rag/utils/pdf.py:
```diff
@@ -0,0 +1,116 @@
+        docs = loader.load()
+        for doc in docs:
+            doc.metadata["source"] = source
+            doc.metadata["coordinates"]["points"] = self._convert_coords(
```
not entirely sure on what the coordinates and points in metadata are for?

sqr00t replied:
These are embeddings (coordinates) within the (HNSW) index space

helloaidank commented on rag/utils/text_processing.py:
```diff
@@ -0,0 +1,44 @@
+from typing import List
+
+
+def nltk_splitter(
```
Another method to try out here would be some semantic chunking and see if it improves the performance.

helloaidank commented on rag/utils/llm.py:
```diff
@@ -0,0 +1,19 @@
+
+from langchain_openai import ChatOpenAI
+
+MODEL_NAME = os.environ.get("MODEL_NAME", "gpt-4o")
```
We changed it back to gpt-3.5-turbo right? I guess that hasn't been pushed to this branch. It would be great to test out `gpt-4o-mini` here!

sqr00t replied:
When you make a request or use langserve remoterunnables, you can add the model name as config. Once I've added the styling on the frontend streamlit app you can test it out that way as well.

sqr00t replied:
Adding a note to update docs on how to use remote runnable with config

helloaidank commented on rag/vector_databases/retrievers.py:
```diff
@@ -0,0 +1,14 @@
+    retriever_searchkwargs_configurable,
+)
+
+chatbot_retriever = (
```
Turn into function, error happens when trying to call vector database locally as still searching for url.

crispy-wonton commented on api/app/utils/events.py:
```diff
@@ -0,0 +1,47 @@
+    # Add any pre-processing logic here.
+
+    # TODO: investigate why logging not working for server deployment. assumption - uvicorn overrides logging
+    logger = logging.getLogger("events.lifespan")
```
How come the params are in quotes here even though it looks like you're calling a variable?

crispy-wonton commented on api/build.toml:
```diff
@@ -0,0 +1,46 @@
+[tool.poetry]
```
question for my understanding: what do these lines with the square brackets do?

crispy-wonton commented on api/app/client.py:
Given that I and likely others on the team aren't that familiar with app code, would it be possible to add 1-2 lines of docstring at the beginning of each of the .py files in the app/ and app/utils/ directories just to summarise what the file does?

crispy-wonton commented on api/build.toml:
```diff
@@ -0,0 +1,46 @@
+langchain-core = "0.2.4"
+qdrant-client = "1.8.2"
+langfuse = "2.26.1"
+torch = {url = "https://download.pytorch.org/whl/cpu/torch-2.3.1%2Bcpu-cp310-cp310-linux_x86_64.whl"}
+torchvision = {url = "https://download.pytorch.org/whl/cpu/torchvision-0.18.1%2Bcpu-cp310-cp310-linux_x86_64.whl"}
```
again for my understanding; what's the reason for using specific wheels here?

sqr00t replied:
for the intended build environment, I don't want CUDA capabilities as the instance does not have a GPU. Using cpu-only wheels slims down the built docker image from ~6GB to ~1GB

crispy-wonton commented on api/app/client.py:
```diff
@@ -0,0 +1,60 @@
+    # validator.validate(webhook_url, form_data, twilio_signature)
+    whatsapp_number = form_data["From"]
+    body = form_data["Body"]
+    user_id = get_sender_hash(whatsapp_number)
```
After the whatsapp number is anonymised, it might be worth deleting the original number from the form data for a step to further protect the information.

crispy-wonton commented on api/app/client.py:
```diff
@@ -0,0 +1,60 @@
+    langfuse_handler = langfuse_handler_from_config(
+        trace_name="whatsapp_chatbot_chain",
+        user_id=user_id,
+        session_id=date.today().isoformat(),  # different session_id from chains, langfuse specific
```
What's the reason for a different ID here?

sqr00t replied:
this is a user_id for the Langfuse record

crispy-wonton reviewed (commented):
I reviewed all the files in the `app/` directory. I'm not familiar with the packages used and building apps in general so I just looked for any obvious code errors or ambiguity. I left a couple of suggestions for documenting code and a couple of questions for my understanding.

helloaidank commented on rag/tests/test_retrieval.py:
```diff
@@ -0,0 +1,68 @@
+    logger.info("Chatbot reponse:")
+    logger.info(chatbot_res["answer"])
+    logger.info("Chatbot Sources:")
+    for src in chatbot_res["context"]:
```
at least when I ran this script, this literally loops through a `string` of the `context`, so it's actually just going through all the different characters of a string.

helloaidank replied:
Which is definitely what we don't want here!

sqr00t replied:
I hadn't pushed the bug fix and updated test that we went over the other day. It's pushed now. Please try re-running!

sqr00t committed:
test: update retrieval test

sqr00t committed:
refactor: update chains and retrievers

helloaidank commented on rag/chains.py:
```diff
@@ -0,0 +1,44 @@
+
+from rag.utils.history import get_session_history
+from langchain_core.runnables.history import RunnableWithMessageHistory
+
```
```suggestion
# Define a basic RAG (Retrieval-Augmented Generation) chain
# This chain consists of a prompt template, a language model, and an output parser
```

helloaidank commented on rag/chains.py:
```diff
@@ -0,0 +1,44 @@
+from langchain_core.runnables.history import RunnableWithMessageHistory
+
+rag_chain = chatbot_template | openai_llm | StrOutputParser()
+
```
```suggestion
# Define a RAG chain that includes historical context from previous interactions
```

helloaidank commented:
> # Description
> Revamps the NIBE prototype to use latest Langchain. Also separates responsibilities:
> 
> * asf_hp_installer_chatbot: _experimentation_ of data heavy pipelines, with associated utils
> * rag: all things end-to-end RAG Langchain related, this includes document processing pipelines
> * api: uses Langserve, a FastAPI wrapper. This enables more ways to interact with the chains throughout dev, test, and prod.
> 
> Note: WhatsApp interface hasn't been implemented, and this PR will be updated to reflect these changes. The aim of this PR is to first revamp the prototype to use the latest Langchain LCEL framework.
> 
> Addresses #6 , branch is `2_revamp`
> 
> # Instructions for Reviewer
> ## **Contact me for Langfuse integration details (env vars needed in your `.env` file)**
> 1. Checkout branch `2_revamp`
> 2. Run `make pip-install` if you already have the project conda environment setup
> 3. Test the integration test script that mimics the prototype test script:
> 
> ```shell
> python rag/tests/test_retrieval.py
> ```
> 
> To test the API:
> 
> 1. Have `poetry` installed, feel free to ask me how best to do this
> 2. `conda deactivate`, then `cd api`
> 3. `poetry env use 3.10`, then `poetry install`
> 4. `langchain app serve --app app.server:app`
> 5. Navigate to `0.0.0.0:8000`
> 6. Test out the `/invoke` endpoint by replacing the string in the JSON request `"input": "string"`, with a query.
> 
> Please pay special attention to and other notes
> 
> * Setting up, feel free to ask me questions
> * I'm still rethinking/ implementing the WhatsApp interface and the messaging system, will be best to not serialise the files but to log conversations to a message queuing system i.e. Redis. Then periodically send incremental batches to an S3 parquet file.
> 
> # Checklist:
> * [x]  I have refactored my code out from `notebooks/` N/A
> * [x]  I have checked the code runs
> * [x]  I have tested the code
> * [x]  I have run `pre-commit` and addressed any issues not automatically fixed
> * [x]  I have merged any new changes from `dev`
> * [x]  I have documented the code
>   
>   * [x]  Major functions have docstrings
>   * [x]  Appropriate information has been added to `README`s
> * [x]  I have explained this PR above
> * [x]  I have requested a code review
> 
> # Other additional features to review:
> * [x]  Add CLI for running ingestion and reindexing of docs into vector database
> * [ ]  Add CLI command for saving vector database snapshots to local and desired S3 path
> * [ ]  Add CLI command for recreating an index from snapshot saved locally or in S3 path
> 
> Review instructions, after `make install` or `make pip-install`, test out CLI commands with `chatbot --help`, there are descriptions for other commands available.

Hi Solomon!

Thanks for the really comprehensive work, I am loving the configurability of the chatbot and think this enables us to have a lot of flexibility in terms of what we are able to do. The change to FastAPI from Flask makes sense especially when it comes to asynchronous nature/potential of chatbots and also love the FastAPI documentation that comes with it. 

I was able to eventually get both the test script up and running and then also run the API with a local Qdrant vector database, so that's good. Ran into a couple of bugs which I've commented on and we've discussed (there have been quite a few changes since this PR was put out so that's unsurprising). It would be great to know what the steps are to deploy the chatbot on the EC2 instance, this is potentially we can discuss at some point in the coming weeks. I think it might be useful @Jack-Vines  to review the API part of this PR as myself and @crispy-wonton do not have much experience with these packages or how they set up.  

As a more general comment, I think we are missing some documentation at the top of each file, i.e. a clear message which is readable for those who either have no or some experience with either web applications or RAG pipelines and I think potentially some visual representations would really help us understand the functionality of the different files and the architecture overall. I am happy to help out with this! 

I think as a whole, we also need documentation on how to set everything up from poetry, to LangFuse and of course with regards to running an EC2 instance! It doesn't seem trivial, at least to a beginner like myself, so I think this would be super helpful. 

Thanks again for the stellar work!

helloaidank committed:
add new test script

helloaidank committed:
adding reengineered prompt with new data sources

helloaidank committed:
adding in some changes with regards to testing script ragas and integrating into langfuse

helloaidank committed:
merge conflicts

helloaidank committed:
small change to test_retrieval

helloaidank committed:
including tej materials into the chatbot prompt

sqr00t committed:
deps: remove irrelevant depedencies

sqr00t committed:
refactor(vdb): add options to vdb init, change parallel loading to ProcessPoolExecutor

sqr00t committed:
refactor(embeddings): update embeddings model from ada-002 to 3-small

sqr00t committed:
refactor(llm): update default model from 4o to 4o-mini

sqr00t committed:
docs(chains): add docstrings to rag_with_source chains

sqr00t committed:
WIP: comment out format_source_docs to disable formatting retrieved source documents

sqr00t committed:
WIP: add a Langchain Callback for running Evaluations

sqr00t committed:
refactor(retrievers): enable passing kwargs to retriever

sqr00t committed:
test: update retrieval test to accept kwargs

sqr00t committed:
feat(cli): add test_query command

sqr00t committed:
deps(api): update dependencies, using pydantic>=2

sqr00t committed:
refactor(chains): update api server and whatsapp client to use function generated chains

sqr00t committed:
refactor(api events): disable ngrok tunnelling by default using env vars

sqr00t committed:
feat(evaluation): add endpoint for triggering evaluation scoring pipeline for traces

sqr00t committed:
build(Dockerfile): update build script for best practices

sqr00t committed:
deps: fix uvicorn version conflict

sqr00t committed:
fix(cli): pass click.context to commands

sqr00t committed:
test(retrieval): update retrieval mmr kwargs

sqr00t committed:
WIP: update chatbot system prompt

sqr00t committed:
fix(evaluation pipeline): fix ragas version to 0.1.16 to fix context_utilization metric error

sqr00t committed:
fix(evaluation pipeline): add retreived_contexts, remove context assertion test

sqr00t committed:
fix(vector database): update vdb client initialisation and handling connections

- yields connection objects and handling closing them
 - update interfaces and chains that use vdb client instances

sqr00t committed:
build(deps): freeze ragas version

sqr00t committed:
refactor(evaluation): optimise evaluation run by running as BackgroundTask

- Updates /score endpoint to fire and forget
 - Runs evaluation functions as a BackgroundTask

sqr00t committed:
build(docker compose): update volume mapping and .env file reference
